src/Voron/StorageEnvironment.cs

```diff
@@ -38,8 +38,49 @@ namespace Voron
 {
     public delegate bool UpgraderDelegate(SchemaUpgradeTransactions transactions, int currentVersion, out int versionAfterUpgrade);
 
-    public class StorageEnvironment : IDisposable
+
+    public class StorageEnvironmentSynchronization
     {
+        public struct SynchronizationSection : IDisposable
+        {
+            private readonly Semaphore _context;
+
+            public SynchronizationSection(Semaphore context)
+            {
+                this._context = context;
+                if (this._context != null)
+                    this._context.WaitOne();
```

Comment:
We need to make sure that this is an _interruptible_ wait.
Right now, if the index is waiting for a write, we can't shut down the server, for example.

Reply:
Makes sense, where do we get that interruption callback from?

Reply:
The easiest is to the use the `DatabaseShutdown` cancellation token. IIRC, there is one for each indexc.